**复习：**在前面我们已经学习了Pandas基础，第二章我们开始进入数据分析的业务部分，在第二章第一节的内容中，我们学习了**数据的清洗**，这一部分十分重要，只有数据变得相对干净，我们之后对数据的分析才可以更有力。而这一节，我们要做的是数据重构，数据重构依旧属于数据理解（准备）的范围。

#### 开始之前，导入numpy、pandas包和数据

In [1]:
# 导入基本库
import numpy as np
import pandas as pd

In [6]:
# 载入上一个任务人保存的文件中:result.csv，并查看这个文件
text=pd.read_csv("result.csv")
text.head()

Unnamed: 0  PassengerId  Survived  Pclass  \
0           0            1         0       3   
1           1            2         1       1   
2           2            3         1       3   
3           3            4         1       1   
4           4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# 2 第二章：数据重构


## 第一部分：数据聚合与运算

### 2.6 数据运用

#### 2.6.1 任务一：通过教材《Python for Data Analysis》P303、Google or anything来学习了解GroupBy机制

##### 写入心得
GroupBy是Pandas提供的强大的数据聚合处理机制，可以对大量级的多维数据进行透视，同时GroupBy还提供强大的apply函数，使得在多维数据中应用复杂函数得到复杂结果成为可能

在了解GroupBy机制之后，运用这个机制完成一系列的操作，来达到我们的目的。

下面通过几个任务来熟悉GroupBy机制。

#### 2.4.2：任务二：计算泰坦尼克号男性与女性的平均票价

In [7]:
# 写入代码
text['Fare']

0       7.2500
1      71.2833
2       7.9250
3      53.1000
4       8.0500
        ...   
886    13.0000
887    30.0000
888    23.4500
889    30.0000
890     7.7500
Name: Fare, Length: 891, dtype: float64

In [10]:
df  = text['Fare'].groupby(text['Sex'])
df.head()

0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
5     8.4583
6    51.8625
7    21.0750
8    11.1333
9    30.0708
Name: Fare, dtype: float64

In [11]:
means = df.mean()
means

Sex
female    44.479818
male      25.523893
Name: Fare, dtype: float64

#### 2.4.3：任务三：统计泰坦尼克号中男女的存活人数

In [12]:
# 写入代码
survived_sex = text['Survived'].groupby(text['Sex']).sum()
survived_sex.head()

Sex
female    233
male      109
Name: Survived, dtype: int64

#### 2.4.4：任务四：计算客舱不同等级的存活人数

In [16]:
# 写入代码
survied_Pclass=text['Survived'].groupby(text['Pclass']).sum()
survied_Pclass.head()

Pclass
1    136
2     87
3    119
Name: Survived, dtype: int64

【**提示：**】表中的存活那一栏，可以发现如果还活着记为1，死亡记为0

【**思考**】从数据分析的角度，上面的统计结果可以得出那些结论

#### 思考心得 
1. 女生平均票价44，男生平均票价25
2. 女生存活人数233，男生存活109
3. 一级客舱存活139，二级存活87，三级存活119
可以看出男女存活的差异，还有贫富的差异。女生应该是贵族的比较多，男生是平民的比较多，然后女生被让出了绝大多数的生还机会。


【思考】从任务二到任务三中，这些运算可以通过agg()函数来同时计算。并且可以使用rename函数修改列名。你可以按照提示写出这个过程吗？

In [18]:
#思考心得
text.groupby('Sex').agg({'Fare': 'mean', 'Pclass': 'count'}).rename(columns=
                            {'Fare': 'mean_fare', 'Pclass': 'count_pclass'})



mean_fare  count_pclass
Sex                            
female  44.479818           314
male    25.523893           577

#### 2.4.5：任务五：统计在不同等级的票中的不同年龄的船票花费的平均值

In [26]:
# 写入代码
text.groupby(['Pclass','Age'])['Fare'].mean()

Pclass  Age  
1       0.92     151.5500
        2.00     151.5500
        4.00      81.8583
        11.00    120.0000
        14.00    120.0000
                   ...   
3       61.00      6.2375
        63.00      9.5875
        65.00      7.7500
        70.50      7.7500
        74.00      7.7750
Name: Fare, Length: 182, dtype: float64

#### 2.4.6：任务六：将任务二和任务三的数据合并，并保存到sex_fare_survived.csv

In [29]:
# 写入代码
means

Sex
female    44.479818
male      25.523893
Name: Fare, dtype: float64

In [30]:
survived_sex

Sex
female    233
male      109
Name: Survived, dtype: int64

In [33]:
result = pd.merge(means,survived_sex,on='Sex')
result.head()

Fare  Survived
Sex                        
female  44.479818       233
male    25.523893       109

#### 2.4.7：任务七：得出不同年龄的总的存活人数，然后找出存活人数最多的年龄段，最后计算存活人数最高的存活率（存活人数/总人数）


In [41]:
# 得出不同年龄的总的存活人数
survived_age=text['Survived'].groupby(text['Age']).sum()
survived_age

Age
0.42     1
0.67     1
0.75     2
0.83     2
0.92     1
        ..
70.00    0
70.50    0
71.00    0
74.00    0
80.00    1
Name: Survived, Length: 88, dtype: int64

In [45]:
# 找出存活人数最多的年龄段
survived_age.max()
survived_age[survived_age.values==survived_age.max()]

Age
24.0    15
Name: Survived, dtype: int64

In [49]:
# 计算存活人数最高的存活率（存活人数/总人数）
# 存活人数
nlive=text['Survived'].sum()
nlive

342

In [50]:
#总人数
nsum = text['Survived'].count()
nsum

891

In [51]:
#总的存活率
precetn=nlive/nsum
precetn

0.3838383838383838

In [55]:
#不同年龄段的人数
all_age=text['Survived'].groupby(text['Age']).count()
all_age

Age
0.42     1
0.67     1
0.75     2
0.83     2
0.92     1
        ..
70.00    2
70.50    1
71.00    2
74.00    1
80.00    1
Name: Survived, Length: 88, dtype: int64

In [60]:
#不同年龄段的存活率
precetn2 = survived_age/all_age
precetn2.head(30)

Age
0.42     1.000000
0.67     1.000000
0.75     1.000000
0.83     1.000000
0.92     1.000000
1.00     0.714286
2.00     0.300000
3.00     0.833333
4.00     0.700000
5.00     1.000000
6.00     0.666667
7.00     0.333333
8.00     0.500000
9.00     0.250000
10.00    0.000000
11.00    0.250000
12.00    1.000000
13.00    1.000000
14.00    0.500000
14.50    0.000000
15.00    0.800000
16.00    0.352941
17.00    0.461538
18.00    0.346154
19.00    0.360000
20.00    0.200000
20.50    0.000000
21.00    0.208333
22.00    0.407407
23.00    0.333333
Name: Survived, dtype: float64

In [62]:
#最大存活率是哪一段年龄？
precetn2[precetn2.values== precetn2.max()]

Age
0.42     1.0
0.67     1.0
0.75     1.0
0.83     1.0
0.92     1.0
5.00     1.0
12.00    1.0
13.00    1.0
53.00    1.0
63.00    1.0
80.00    1.0
Name: Survived, dtype: float64